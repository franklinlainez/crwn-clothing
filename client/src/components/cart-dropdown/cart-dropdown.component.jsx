import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router-dom";
import { toggleCartHidden } from "../../redux/cart/cart.actions";
import { selectCartItems } from "../../redux/cart/cart.selectors";
import CartItem from "../cart-item/cart-item.component";
import {
  CartDropdownButton,
  CartDropdownContainer,
  CartItemsContainer,
  EmptyMessage,
} from "./cart-dropdown.styles";

const CartDropdown = () => {
  const cartItems = useSelector(selectCartItems);
  const dispatch = useDispatch();
  const history = useHistory();

  return (
    <CartDropdownContainer>
      <CartItemsContainer>
        {cartItems.length ? (
          cartItems.map((cartItem) => (
            <CartItem key={cartItem.id} item={cartItem} />
          ))
        ) : (
          <EmptyMessage>Your cart is empty</EmptyMessage>
        )}
      </CartItemsContainer>
      <CartDropdownButton
        onClick={() => {
          history.push("/checkout");
          dispatch(toggleCartHidden());
        }}
      >
        GO TO CHECKOUT
      </CartDropdownButton>
    </CartDropdownContainer>
  );
};

export default CartDropdown;
