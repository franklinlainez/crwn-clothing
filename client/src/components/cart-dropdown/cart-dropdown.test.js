import { mount, ReactWrapper } from "enzyme";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { toggleCartHidden } from "../../redux/cart/cart.actions";
import CartItem from "../cart-item/cart-item.component";
import CartDropdown from "./cart-dropdown.component";
import { CartDropdownButton, EmptyMessage } from "./cart-dropdown.styles";

const mockStore = configureMockStore([]);
const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

const mockHistoryPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe("CartDropdown component", () => {
  /**@type ReactWrapper */
  let wrapper;

  let store = mockStore({});

  beforeEach(() => {
    store = mockStore({
      cart: {
        cartItems: mockCartItems,
      },
    });

    wrapper = mount(
      <Provider store={store}>
        <CartDropdown />
      </Provider>
    );
  });

  it("expects a render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call history.push when button is clicked", () => {
    wrapper.find(CartDropdownButton).simulate("click");
    expect(mockHistoryPush).toHaveBeenCalled();
    expect(store.getActions()).toContainEqual(toggleCartHidden());
  });

  it("should render an equal number of CartItem components as the cartItems prop", () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });

  it("should render EmptyMessageContainer if cartItems is empty", () => {
    const newWrapper = mount(
      <Provider
        store={mockStore({
          cart: {
            cartItems: [],
          },
        })}
      >
        <CartDropdown />
      </Provider>
    );

    expect(newWrapper.find(EmptyMessage).length).toBe(1);
  });
});
