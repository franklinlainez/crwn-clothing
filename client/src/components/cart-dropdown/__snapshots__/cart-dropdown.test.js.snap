// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component expects a render 1`] = `ReactWrapper {}`;
